src/riot-rs-macros/src/main_macro.rs

```diff
@@ -0,0 +1,200 @@
+/// Registers an async function acting as an entrypoint for an application.
+///
+/// The function is provided with a peripheral struct as the first parameter, if present.
+/// The peripheral struct must be defined with the `define_peripherals!` macro.
+///
+/// **Important**: The `embassy_executor` crate currently needs to be manually imported in the crate using this
+/// attribute macro.
+///
+/// If this function is only used to spawn other tasks before returning, consider using
+/// [`macro@spawner`] instead, to avoid statically allocating this transient async function as an
+/// `embassy_executor::task`.
+///
+/// # Parameters
+///
+/// - hooks: (*optional*) list of hooks. Available hooks are:
+///     - `usb_builder_hook`: when present, the macro will define a static `USB_BUILDER_HOOK` of
+///     type `UsbBuilderHook`, allowing to access and modify the system-provided
+///     `embassy_usb::Builder` through `Delegate::with()`, *before* it is built by the system.
+///
+/// # Examples
+///
+/// ```ignore
+/// use riot_rs::embassy::usb::UsbBuilderHook;
+///
+/// #[riot_rs::main(usb_builder_hook)]
+/// async fn main(peripherals: /* your peripheral type */) {}
+/// ```
+///
+/// See RIOT-rs examples for more.
+///
+/// # Panics
+///
+/// This macro panics when the `riot-rs` crate cannot be found as a dependency of the crate where
+/// this macro is used.
+#[proc_macro_attribute]
+pub fn main(args: TokenStream, item: TokenStream) -> TokenStream {
+    use quote::quote;
+
+    #[allow(clippy::wildcard_imports)]
+    use main_macro::*;
+
+    let mut attrs = MainAttributes::default();
+    let main_attr_parser = syn::meta::parser(|meta| attrs.parse(&meta));
+    syn::parse_macro_input!(args with main_attr_parser);
+
+    let main_function = syn::parse_macro_input!(item as syn::ItemFn);
+    let main_function_name = &main_function.sig.ident;
+    let is_async = main_function.sig.asyncness.is_some();
+
+    assert!(is_async, "the function must be async");
+
+    let riot_rs_crate = utils::riot_rs_crate();
+
+    let hooks = Hook::hook_definitions();
+
+    let takes_peripherals = !main_function.sig.inputs.is_empty();
+    let peripheral_param = if takes_peripherals {
+        quote! {peripherals.take_peripherals()}
+    } else {
+        quote! {}
+    };
+
+    let delegates = main_macro::generate_delegates(&riot_rs_crate, &hooks, &attrs);
+
+    let expanded = quote! {
+            #delegates
+
+            #[#riot_rs_crate::embassy::distributed_slice(#riot_rs_crate::embassy::EMBASSY_TASKS)]
+            #[linkme(crate = #riot_rs_crate::embassy::linkme)]
+            fn __main(
```

Comment:
call it  `__start_#main_function_name()`?

IIUC like this, it fails if there is more than one `#[riot_rs::main]`?

Reply:
> and, if they are using hooks, this is very likely to cause deadlocks.

I don't think that's accurate. The hooks need to be used in the order that `init_tasks()` is await'ing them. multiple tasks using hooks are no issue other than name clashes, which we can work around in our macro. (otherwise, multiple tasks in different files would pose a problem as well).

Reply:
> It would indeed fail if `#[riot_rs::main]` is used on multiple functions in the same module;

This is btw a clearly desired feature, if not just for the examples and tests and doc tests.

```
/// ... riot-rs boilerplate

static CHANNEL: ... = SomeChannel::new();
#[riot_rs::task]
fn task0() {
  CHANNEL.send(42);
}
#[riot_rs::task]
fn task1() {
  assert_eq!(42, CHANNEL.recv());
}

Reply:
> I think `riot_rs::task` should just be the re-export of `embassy_executor::task`

We could add an "autostart" parameter (which defaults to true). `#[embassy_executor::task]` would then become `#[riot_rs::task(autostart=false)]`. (At that point `task` is not a re-export anymore though, and the `pool` parameter would need to get passed through.)

I'm thinking, there is already `#[riot_rs::thread]`, it would be cool if their APIs would be somewhat similar.

Reply:
> The issue with that is that the function signature would be completely different

Ah, of course. :/

Reply:
So what's more important:

1. consistency between the concepts `task` and `thread`
2. keeping `riot_rs::task` as reexport of `embassy_executor::task`

I tend to the former. I'd even be fine with keeping the use of `embassy_executor::task` until we've found a better name.

Let's get more opinions!